### PR TITLE
Asynchronous Contents API

### PR DESCRIPTION
**Pull request by mwakaba2**
**Description**
Hi, here's the draft for adding an async option for the contents api. 

So far, I added an async option for each class/mixin below:
- `ContentsManager` and `AsyncContentsManager`
- `Checkpoints` and `AsyncCheckpoints`
- `GenericCheckpointsMixin` and `AsyncGenericCheckpointsMixin`

@kevin-bates and/or @Zsailer I'd appreciate your review and feedback!

**Open Questions**
I added a test for the configuration, but how do we want to test the new classes since most methods are not implemented and left up to the user?

**Checklist**
- [x] Update Contents Api Doc
- [x] Add test for contents configuration
- [x] Implement `AsyncFileContentsManager`
- [x] Update test_ap/test_manager to test both `FileContentsManager` and `AsyncFileContentsManager`
- [x] Rebase with master branch to include public fixture update. 
- [x] Remove python 3.5 support code. 
- [x] Fix code scanning lint errors
- [x] Test new classes with my team's custom `ContentsManager`
- [x] Update docs to include best use cases for the async classes. 
- [x] Fix tests

### CODE REVIEW

**Comment by kevin-bates** [comment]
Thanks @mwakaba2 - this is a good start.

> how do we want to test the new classes since most methods are not implemented and left up to the user?

You're right, most tests leverage the `FileContentsManager` and things should get "concrete" when adding its async form.  At that time, I suspect we could use parameters like in the kernels tests, which run the tests using [both classes](https://github.com/jupyter/jupyter_server/blob/master/tests/services/kernels/test_api.py#L18-L22)...
```
@pytest.fixture(params=["MappingKernelManager", "AsyncMappingKernelManager"])
def argv(request):
    if request.param == "AsyncMappingKernelManager" and sys.version_info < (3, 6):
        pytest.skip("Kernel manager is AsyncMappingKernelManager, Python version < 3.6")
    return ["--ServerApp.kernel_manager_class=jupyter_server.services.kernels.kernelmanager." + request.param]
```

**Comment by mwakaba2** [comment]
@kevin-bates Since `FileContentsManager` is a subclass of `ContentsManager`, one option is to create  `AsyncFileContentsManager` for asynchronous file operations and test that. Then, I think we'll be able to test both classes in the test_api file. However, it looks like asyncio doesn't support async file operations by default, so we'll need to use something like [aiofiles: file support for asyncio](https://github.com/Tinche/aiofiles). 

Another option would be to create a no-op or simple custom ContentsManager class that inherits from the async classes, and test that. To mimic network IO, the async methods can include `asyncio.sleep` statements.
I think the first option would be a complete solution and can rely on the existing tests, whereas the second will not exactly test real situations. What would you recommend?

**Comment by kevin-bates** [comment]
Hi Mariko - I think taking the _forward-moving_ branch by implementing `AsyncFileContentsManager` is the best approach.  Thank you.

**Comment by mwakaba2** [comment]
Hi @kevin-bates, I added the `AsyncFileContentsManager` class and updated tests to test both async and non-async classes. 
Instead of `aiofiles`, I added a new dependency [`anyio`](https://anyio.readthedocs.io/en/latest/index.html) to run some synchronous callables in a thread with [`run_sync_in_worker_thread`](https://anyio.readthedocs.io/en/latest/threads.html#running-a-function-in-a-worker-thread). This is used to run most `os` operations and `nbformat` methods concurrently. The nbformat team is looking into making the api asynchronous (https://github.com/jupyter/nbformat/issues/194), but it looks like that'll take some time to get that feature out, so using `anyio` for now. 
Please let me know if you have any questions and I'd appreciate your review!

File: tests/services/contents/test_api.py
Comment (kevin-bates) [inline tests/services/contents/test_api.py:48]: Copy/paste - `Kernel` -> `Contents`.  Actually, since we no longer support 3.5, this condition can be dropped altogether. :smile:

**Comment by kevin-bates** [comment]
Hi @mwakaba2 - thank you for this great update.  I hope to be able to take this for spin next week - I hope that's okay.  Things look real good via a quick scan - just had the one immediate comment (to show I did look at things :smile: ).

Also, we've made some fairly large changes to the pytest landscape.  This will manifest in your additional tests in the form for requiring `fp_` prefix on the "public" fixtures - which I suspect to be fairly isolated, but a merge with master might be helpful.

Thanks for the links to `nbformat`. The additional dependency on `anyio` seems fine, especially since @bollwyvl is on board!

**Comment by mwakaba2** [comment]
Hi @kevin-bates, hope you had a good weekend. Yeah that works! I see some code scanning lint errors, so I'll fix those along with removing the python 3.5 pytest skip logic, rebasing with master to include public fixture changes, and testing the new classes with my team's custom contentsmanager. Thanks for the quick review!

**Comment by mwakaba2** [comment]
Hi @kevin-bates, is it possible to publish a dev version to PyPI to test this branch?

**Comment by kevin-bates** [comment]
Hi @mwakaba2 - I'm not sure there's precedent for this and reading up on publishing dev packages to pypi (as I'm not familiar with that particular aspect) I'm finding it not a recommended practice.

What I've done in the past is to use pip with a reference to the GH repo@branch, like so:
```
pip install --upgrade git+https://github.com/mwakaba2/jupyter_server@async-contents-api
```
Would that be sufficient?

Btw, I brought up your PR in the today's server meeting and will try to spend time with this.  I apologize for the delay.

**Comment by mwakaba2** [comment]
Ah ok no problem. Yeah, we're using an in-house tool to manage 3rd party python dependencies so I'm not able to use pip directly, but I think I found a way to test it.
And thanks! I saw the meeting notes. I'm hoping to complete testing this week.

**Comment by kevin-bates** [comment]
Hi Mariko - I was just going to post that `LargeFileManager` should be included since it's currently the default `contents_manager_class` value.  Thank you for the update.

I was able to take `AsyncFileContentsManager` for a spin yesterday and found the functionality to be fine.  I decided to run some _crude_ tests using `curl` and uploading/fetching [larger file contents (189 MB)](https://github.com/zemirco/sf-city-lots-json/blob/master/citylots.json).  A "test" simply consisted of timing 3 curl commands (each in the background) to simulate 3 "simultaneous" requests.  Larger content was used solely to trigger longer request times and ensure the previous requests hadn't finished prior to the subsequent requests.  I.e., I wasn't as concerned about the request time as I was about getting multiple requests "serviced" simultaneously.

I found that the current implementation didn't exhibit the blocking behavior that I experienced with kernel startup before we implemented async/await support - which surprised me.  I also found that it seems like the current implementation was slightly better in that you could see a "clear winner" amongst the 3 requests (upwards of 50% faster), whereas, with the async implementation, the times tended to be the same across the 3 requests.  I suspect this is because of "too much cooperation" since each `await` is essentially a _pause point_ for that executor.

Since the primary purpose of this is to address high-latent filesystems, I suspect there will be some "wins" there.  I also wonder if we should take a close look at where `awaits` have been introduced and ensure they make sense.

I'm curious if you see similar behavior.  Thank you for working on this.

**Comment by kevin-bates** [comment]
Hi Mariko, would you mind splitting the PEP440 update out in a separate PR?  This is something I had on my plate to do today (along with an update to [RELEASE.md](https://github.com/jupyter-server/jupyter_server/blob/master/RELEASE.md) to update the dev version following the release tagging).

I think only the `dev` and `post` suffixes use a dot: https://www.python.org/dev/peps/pep-0440/#public-version-identifiers

If you'd rather I do that, that's fine also.

**Comment by kevin-bates** [comment]
:smile:  Apparently someone also had this on their plate _yesterday_: https://github.com/jupyter-server/jupyter_server/commit/8763b5b98cb745cc11713af20f72a1919ea2d181

I'll go ahead and submit a PR to add the dot and update RELEASE.md.

**Comment by jasongrout** [comment]
> I'll go ahead and submit a PR to add the dot and update RELEASE.md.

You may remember that we did this sort of thing in https://github.com/jupyter/jupyter_core/blob/master/jupyter_core/version.py

**Comment by kevin-bates** [comment]
Thanks @jasongrout - I _just_ saw your comment (after submitting #348).  I hadn't forgotten about your change in core, but personally found that to be overly complex.  That said, if others prefer that approach, I'm okay with adopting it.

**Comment by jasongrout** [comment]
> personally found that to be overly complex

It could have been way more complex for full pep440 compliance, but we dialed it back :).

The purpose for the extra stuff in that file is to provide a good versioninfo tuple that mirrors the versioninfo tuple from Python, and to make it so that you don't have to think about how pep440 formats version numbers (dot? consistent abbreviation?). I think it probably does that as simply as possible, but (a) if there is something simpler, I'd love to see it - I think it's a bit much to stomach as well, and (b) of course, if you don't want those two goals, that's a different story.

**Comment by mwakaba2** [comment]
Hi @kevin-bates, I have some bugs to fix, so I think I need a little more time to test this out, but yep I agree. There are definitely some `await`s that can be removed. I'm interested in the tests you performed and would like to try it out after I fix some bugs on my end. Could you share the commands you ran for the `curl` tests?

As for the pep440 update, I made that quick change just to see if the tests are still working in this branch. I'll remove the pep440 commit and update this branch once your PR is merged to the main branch. 

I'm taking next week off, so I'll be able to resume testing the week after that (11/30). I noticed you're taking time off in December and was wondering if another jupyter member can help with the code review/release?

**Comment by kevin-bates** [comment]
Thanks @mwakaba2 - that all sounds great.  Have a great week off!  

Although I'll be out for (essentially) December, my plan is to be around for this kind of stuff.  @Zsailer can you help review this and move this forward as well.  Perhaps we can target the Server Team meeting of December 3 to sync up with things?

**Comment by kevin-bates** [comment]
I just realized you had asked to see the test scripts. I've included them in the attached tar file but did not include the large content file. The README describes how to set these up and how to construct the larger payload.  As I said, they are crude and minimally viable. :smile:

[contents-test.tar.gz](https://github.com/jupyter-server/jupyter_server/files/5577989/contents-test.tar.gz)

**Comment by mwakaba2** [comment]
@kevin-bates Thanks for putting together the test bundle. 
I ran your put/get scripts 5 times to calculate the avg for each scenario:
  * Before (current implementation on master branch)
```
$ jupyter server --ServerApp.contents_manager_class=jupyter_server.services.contents.largefilemanager.FileContentsManager --debug on main branch)
```
  * After (FileContentsManager)
```
$ jupyter server --ServerApp.contents_manager_class=jupyter_server.services.contents.largefilemanager.FileContentsManager --debug
```
  * After (AsyncFileContentsManager)
```
$ jupyter server --ServerApp.contents_manager_class=jupyter_server.services.contents.largefilemanager.AsyncFileContentsManager --debug
```
Here are the results. 
Fastest request out of three "huge" requests
```
Before
---------
PUT
201 PUT /api/contents/contents-test/huge1.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 604.49ms
201 PUT /api/contents/contents-test/huge2.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 517.84ms
201 PUT /api/contents/contents-test/huge2.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 555.70ms
201 PUT /api/contents/contents-test/huge2.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 490.21ms
201 PUT /api/contents/contents-test/huge2.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 707.03ms
---------
avg 575.054ms

GET
200 GET /api/contents/contents-test/huge2.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 941.87ms
200 GET /api/contents/contents-test/huge2.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 843.39ms
200 GET /api/contents/contents-test/huge1.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 873.16ms
201 PUT /api/contents/contents-test/huge1.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 506.65ms
200 GET /api/contents/contents-test/huge2.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 873.18ms
---------
avg 807.65ms

After (non-async)
-----------------
PUT
201 PUT /api/contents/contents-test/huge1.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 727.68ms
201 PUT /api/contents/contents-test/huge2.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 520.14ms
201 PUT /api/contents/contents-test/huge1.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 494.54ms
201 PUT /api/contents/contents-test/huge2.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 558.08ms
201 PUT /api/contents/contents-test/huge3.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 599.44ms
---------
avg 579.976ms

GET
200 GET /api/contents/contents-test/huge1.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 836.26ms
200 GET /api/contents/contents-test/huge1.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 819.93ms
200 GET /api/contents/contents-test/huge2.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 834.84ms
200 GET /api/contents/contents-test/huge2.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 857.20ms
200 GET /api/contents/contents-test/huge2.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 828.08ms
---------
avg 835.26ms

After (async)
-----------------
PUT
201 PUT /api/contents/contents-test/huge1.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 1253.11ms
201 PUT /api/contents/contents-test/huge2.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 893.71ms
201 PUT /api/contents/contents-test/huge3.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 917.52ms
201 PUT /api/contents/contents-test/huge1.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 1457.99ms
201 PUT /api/contents/contents-test/huge2.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 1363.17ms
---------
avg 1177.1ms

GET
200 GET /api/contents/contents-test/huge1.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 859.31ms
200 GET /api/contents/contents-test/huge1.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 851.68ms
200 GET /api/contents/contents-test/huge2.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 876.50ms
200 GET /api/contents/contents-test/huge2.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 886.85ms
200 GET /api/contents/contents-test/huge3.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 859.66ms
---------
avg 866.8ms
```
I confirmed that there's no noticeable performance difference between "Before" and "After (FileContentsManager)", so that's good. 😀
As for the asynchronous implementation, as you saw, it looks like there's a 104.69% increase in response time for the fastest "huge" PUT request. The increase is minimal for the GET request.

So I removed some await statements to reduce the PUT response time. The increase is minimal now.
```
After (async, optimal)
-----------------
PUT
201 PUT /api/contents/contents-test/huge1.json?token=4f24b1be4ce39fdfd8d911b737f718e1572e4c1bc2dd7c93 (::1) 747.23ms
201 PUT /api/contents/contents-test/huge1.json?token=4f24b1be4ce39fdfd8d911b737f718e1572e4c1bc2dd7c93 (::1) 512.65ms
201 PUT /api/contents/contents-test/huge2.json?token=4f24b1be4ce39fdfd8d911b737f718e1572e4c1bc2dd7c93 (::1) 587.67ms
201 PUT /api/contents/contents-test/huge2.json?token=4f24b1be4ce39fdfd8d911b737f718e1572e4c1bc2dd7c93 (::1) 576.43ms
201 PUT /api/contents/contents-test/huge3.json?token=4f24b1be4ce39fdfd8d911b737f718e1572e4c1bc2dd7c93 (::1) 538.99ms
---------
avg 592.59ms
```
This is just an update for PUT, so I'll look into optimizations for the other file operations next (e.g. rename, delete).

**Comment by jhamet93** [comment]
> @kevin-bates Thanks for putting together the test bundle.
> I ran your put/get scripts 5 times to calculate the avg for each scenario:
> 
> * Before (current implementation on master branch)
> 
> ```
> $ jupyter server --ServerApp.contents_manager_class=jupyter_server.services.contents.largefilemanager.FileContentsManager --debug on main branch)
> ```
> 
> * After (FileContentsManager)
> 
> ```
> $ jupyter server --ServerApp.contents_manager_class=jupyter_server.services.contents.largefilemanager.FileContentsManager --debug
> ```
> 
> * After (AsyncFileContentsManager)
> 
> ```
> $ jupyter server --ServerApp.contents_manager_class=jupyter_server.services.contents.largefilemanager.AsyncFileContentsManager --debug
> ```
> 
> Here are the results.
> Fastest request out of three "huge" requests
> 
> ```
> Before
> ---------
> PUT
> 201 PUT /api/contents/contents-test/huge1.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 604.49ms
> 201 PUT /api/contents/contents-test/huge2.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 517.84ms
> 201 PUT /api/contents/contents-test/huge2.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 555.70ms
> 201 PUT /api/contents/contents-test/huge2.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 490.21ms
> 201 PUT /api/contents/contents-test/huge2.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 707.03ms
> ---------
> avg 575.054ms
> 
> GET
> 200 GET /api/contents/contents-test/huge2.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 941.87ms
> 200 GET /api/contents/contents-test/huge2.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 843.39ms
> 200 GET /api/contents/contents-test/huge1.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 873.16ms
> 201 PUT /api/contents/contents-test/huge1.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 506.65ms
> 200 GET /api/contents/contents-test/huge2.json?token=fd559df06790466297db4a89b210b66ef79a104e935ba17d (::1) 873.18ms
> ---------
> avg 807.65ms
> 
> After (non-async)
> -----------------
> PUT
> 201 PUT /api/contents/contents-test/huge1.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 727.68ms
> 201 PUT /api/contents/contents-test/huge2.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 520.14ms
> 201 PUT /api/contents/contents-test/huge1.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 494.54ms
> 201 PUT /api/contents/contents-test/huge2.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 558.08ms
> 201 PUT /api/contents/contents-test/huge3.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 599.44ms
> ---------
> avg 579.976ms
> 
> GET
> 200 GET /api/contents/contents-test/huge1.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 836.26ms
> 200 GET /api/contents/contents-test/huge1.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 819.93ms
> 200 GET /api/contents/contents-test/huge2.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 834.84ms
> 200 GET /api/contents/contents-test/huge2.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 857.20ms
> 200 GET /api/contents/contents-test/huge2.json?token=b9c8c67f686f6b746cc4479ddbaee06f8526659d0b10b5a1 (::1) 828.08ms
> ---------
> avg 835.26ms
> 
> After (async)
> -----------------
> PUT
> 201 PUT /api/contents/contents-test/huge1.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 1253.11ms
> 201 PUT /api/contents/contents-test/huge2.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 893.71ms
> 201 PUT /api/contents/contents-test/huge3.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 917.52ms
> 201 PUT /api/contents/contents-test/huge1.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 1457.99ms
> 201 PUT /api/contents/contents-test/huge2.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 1363.17ms
> ---------
> avg 1177.1ms
> 
> GET
> 200 GET /api/contents/contents-test/huge1.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 859.31ms
> 200 GET /api/contents/contents-test/huge1.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 851.68ms
> 200 GET /api/contents/contents-test/huge2.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 876.50ms
> 200 GET /api/contents/contents-test/huge2.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 886.85ms
> 200 GET /api/contents/contents-test/huge3.json?token=207ea8b331c25affbb8e607138e46fff0c587935d83cd3f4 (::1) 859.66ms
> ---------
> avg 866.8ms
> ```
> 
> I confirmed that there's no noticeable performance difference between "Before" and "After (FileContentsManager)", so that's good. 😀
> As for the asynchronous implementation, as you saw, it looks like there's a 104.69% increase in response time for the fastest "huge" PUT request. The increase is minimal for the GET request.
> 
> So I removed some await statements to reduce the PUT response time. The increase is minimal now.
> 
> ```
> After (async, optimal)
> -----------------
> PUT
> 201 PUT /api/contents/contents-test/huge1.json?token=4f24b1be4ce39fdfd8d911b737f718e1572e4c1bc2dd7c93 (::1) 747.23ms
> 201 PUT /api/contents/contents-test/huge1.json?token=4f24b1be4ce39fdfd8d911b737f718e1572e4c1bc2dd7c93 (::1) 512.65ms
> 201 PUT /api/contents/contents-test/huge2.json?token=4f24b1be4ce39fdfd8d911b737f718e1572e4c1bc2dd7c93 (::1) 587.67ms
> 201 PUT /api/contents/contents-test/huge2.json?token=4f24b1be4ce39fdfd8d911b737f718e1572e4c1bc2dd7c93 (::1) 576.43ms
> 201 PUT /api/contents/contents-test/huge3.json?token=4f24b1be4ce39fdfd8d911b737f718e1572e4c1bc2dd7c93 (::1) 538.99ms
> ---------
> avg 592.59ms
> ```
> 
> This is just an update for PUT, so I'll look into optimizations for the other file operations next (e.g. rename, delete).

I'm not sure having matching performance parity with `LargeFileContentsManager` is necessarily the target to shoot for so I would just like to caution against removing `await` in an effort to close the gap. Tasks that are I/O bound are well suited for `await` so if there are any that do not fit that bill by all means remove them. 

 It should just be well documented that this solution is best for latent filesystems such as a VFS operating over the network which can take advantage of the performance gains ?  Thus would be very curious on that front the performance differences this change results in.

**Comment by mwakaba2** [comment]
yeah you're right. I reviewed all await statements and they are all IO related, so I'm not sure if I can make any optimizations. For example, to improve the performance for uploading files, I just reverted the change of making `_save_file` asynchronous which kind of defeats the purpose of making the `FileContentsManager/LargeFileManager` async in the first place . 
I can definitely update the docs to include a note about best use cases for the async classes. 
I'll also share any performance diffs using the async classes in a VFS.

**Comment by mwakaba2** [comment]
Hi @kevin-bates and @Zsailer, the PR is ready for review.
After testing my team’s custom manager with the asynchronous Contents API, I confirmed that the terminal lag was reduced by **12 ~ 38%.**

I performed three tests:
1. Executing a ~10 second command in the terminal with no user-triggered notebook operations: **12.57%** decrease in command execution time. (the files are fetched in the background every 10 seconds.)
2. Executing a ~10 second command while manually fetching a large notebook (1.1 MB): **28.9%** decrease in command execution time.
3. Executing a ~10 second command while saving a large notebook (1.1 MB): **37.97%** decrease in command execution time
The percentage was calculated by comparing before and after results. (Result = average of 5 command execution times)

**Review by kevin-bates — APPROVED** [review]
This is a great contribution @mwakaba2.  As Jupyter expands across more filesystem types and concurrency demands increase, I believe the async contents you've provided will prove useful - thank you!

File: jupyter_server/services/contents/fileio.py
Comment (Zsailer) [inline jupyter_server/services/contents/fileio.py:205]:
```suggestion
        """wrapper around io.open that turns permission errors into 403"""
```

**Review by Zsailer — APPROVED** [review]

File: tests/services/contents/test_manager.py
Comment (Zsailer) [inline tests/services/contents/test_manager.py:27]: This should be prefixed with `jp_`, but let's not block on this.

**Comment by Zsailer** [comment]
Thanks, @mwakaba2! Looks great! Merging away!

**Pull request merged**
